frame/two-phase-election-provider/src/lib.rs

```diff
@@ -0,0 +1,1627 @@
+	/// 3. for each assignment, the check of `ElectionDataProvider` is also examined.
+	/// 4. the claimed score is valid.
+	fn feasibility_check(
+		solution: RawSolution<CompactOf<T>>,
+		compute: ElectionCompute,
+	) -> Result<ReadySolution<T::AccountId>, FeasibilityError> {
+		let RawSolution {
+			compact,
+			score,
+			round,
+		} = solution;
+
+		// first, check round.
+		ensure!(Self::round() == round, FeasibilityError::InvalidRound);
+
+		// winners are not directly encoded in the solution.
+		let winners = compact.unique_targets();
+
+		// TODO: this can probably checked ahead of time, both in signed and unsigned story.
+		let desired_targets =
+			Self::desired_targets().ok_or(FeasibilityError::SnapshotUnavailable)?;
+
+		ensure!(
+			winners.len() as u32 == desired_targets,
+			FeasibilityError::WrongWinnerCount,
+		);
+
+		// read the entire snapshot.
+		let RoundSnapshot {
+			voters: snapshot_voters,
+			targets: snapshot_targets,
+		} = Self::snapshot().ok_or(FeasibilityError::SnapshotUnavailable)?;
+
+		// ----- Start building. First, we need some closures.
+		crate::voter_at_fn!(let voter_at , snapshot_voters, T);
+		crate::target_at_fn!(let target_at, snapshot_targets, T);
+		crate::voter_index_fn_usize!(let voter_index, snapshot_voters, T);
+
+		// first, make sure that all the winners are sane.
+		let winners = winners
+			.into_iter()
+			.map(|i| target_at(i).ok_or(FeasibilityError::InvalidWinner))
+			.collect::<Result<Vec<T::AccountId>, FeasibilityError>>()?;
+
+		// Then convert compact -> Assignment. This will fail if any of the indices are gibberish.
+		// that winner indices are already checked.
+		let assignments = compact
+			.into_assignment(voter_at, target_at)
+			.map_err::<FeasibilityError, _>(Into::into)?;
+
+		// Ensure that assignments is correct.
+		let _ = assignments
+			.iter()
+			.map(|ref assignment| {
+				// check that assignment.who is actually a voter (defensive-only).
+				// NOTE: while using the index map from `voter_index` is better than a blind linear
+				// search, this *still* has room for optimization. Note that we had the index when
+				// we did `compact -> assignment` and we lost it. Ideal is to keep the index around.
+
+				// defensive-only: must exist in the snapshot.
+				let snapshot_index = voter_index(&assignment.who).ok_or(FeasibilityError::InvalidVoter)?;
+				// defensive-only: index comes from the snapshot, must exist.
+				let (_voter, _stake, targets) = snapshot_voters.get(snapshot_index).ok_or(FeasibilityError::InvalidVoter)?;
+
+				// check that all of the targets are valid based on the snapshot.
+				if assignment
+					.distribution
+					.iter()
+					.any(|(d, _)| !targets.contains(d))
+				{
+					return Err(FeasibilityError::InvalidVote);
+				}
+				Ok(())
+			})
+			.collect::<Result<(), FeasibilityError>>()?;
+
+		// ----- Start building support. First, we need some more closures.
+		stake_of_fn!(let stake_of, snapshot_voters, T);
```

Comment:
why do we create a stake_map inside this ? can't we just reuse the `voters_map` created in `voter_index_fn_usize` ?
building it 2 times seems a waste of ressource, no ?

Reply:
Indeed, fixed it.